engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala

```diff
@@ -7549,38 +7548,19 @@ object IR {
         *
         * @param err the original error.
         */
-      case class ResolverError(err: BindingsMap.ResolutionError)
+      case class ResolverError(private val explain: ExplainResolution)
           extends Reason {
 
         /** Provides a human-readable explanation of the error.
           * @param originalName the original unresolved name.
           * @return a human-readable message.
           */
         override def explain(originalName: IR.Name): String =
-          err match {
-            case BindingsMap.ResolutionAmbiguous(candidates) =>
-              val firstLine =
-                s"The name ${originalName.name} is ambiguous. Possible candidates are:"
-              val lines = candidates.map {
-                case BindingsMap.ResolvedConstructor(
-                      definitionType,
-                      cons
-                    ) =>
-                  s"    Constructor ${cons.name} defined in module ${definitionType.module.getName};"
-                case BindingsMap.ResolvedModule(module) =>
-                  s"    The module ${module.getName};"
-                case BindingsMap.ResolvedPolyglotSymbol(_, symbol) =>
-                  s"    The imported polyglot symbol ${symbol.name};"
-                case BindingsMap.ResolvedMethod(module, symbol) =>
-                  s"    The method ${symbol.name} defined in module ${module.getName}"
-                case BindingsMap.ResolvedType(module, typ) =>
-                  s"    Type ${typ.name} defined in module ${module.getName}"
-              }
-              (firstLine :: lines).mkString("\n")
-            case BindingsMap.ResolutionNotFound =>
-              s"The name `${originalName.name}` could not be found"
-          }
+          this.explain.explain(originalName)
+      }
 
+      trait ExplainResolution {
```

Comment:
Reverts the _direction of dependency_. Rather than `ResolverError` depending on `BindingsMap`, it exposes an interface that `BindingsMap` then implements.